contracts/DuskMainnetOnramp.sol

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: MPL-2.0
+pragma solidity ^0.8.24;
+
+// Minimal ERC20 interface required to transfer DUSK
+interface IERC20 {
+    function transferFrom(
+        address from,
+        address to,
+        uint256 value
+    ) external returns (bool);
+}
+
+/// @title The ERC20/BEP20 DUSK mainnet on-ramp contract
+/// @author Hein Dauven
+contract DuskMainnetOnramp {
+    IERC20 public immutable duskToken;
+    // Conversion factor between ERC20 DUSK (18 decimals) and native DUSK (9 decimals), where 10^9 DUSK wei is equivalent to 1 LUX
+    uint256 constant LUX_CONVERSION_FACTOR = 10**9; 
+    uint256 public endTime;
+
+    // Event to log the genesis deposit for inclusion in the genesis state on Dusk mainnet. The amount being in LUX
+    event GenesisDeposit(address indexed from, uint256 amount, string targetAddress);
+
+    /**
+     * @param _duskTokenAddress The address of the ERC20/BEP20 DUSK token contract.
+     * @param _endTime The time after which deposits are no longer allowed.
+     */
+    constructor(address _duskTokenAddress, uint256 _endTime) {
+        duskToken = IERC20(_duskTokenAddress);
+        endTime = _endTime;
+    }
+
+    /**
+     * @notice Moves ERC20 DUSK tokens to native DUSK by transferring the tokens from the sender to this contract for locking.
+     *         The function rounds the amount down to the nearest 1 LUX (10^9 DUSK wei).
+     * 
+     * @dev This function follows a simple check-interactions pattern to minimize reentrancy risk:
+     *      1. Check: Check if the `amount` is greater than or equal to 1 LUX.
+     *      2. Interaction: Transfers the specified amount of DUSK tokens to the contract and emits a `GenesisDeposit` event.
+     * 
+     * @dev We assume that the targetAddress is a valid Moonlight key. The user will never explicitly type in the key, it will instead be provided by the UX.
```

Comment:
```suggestion
     * @dev We assume that the targetAddress is a valid Moonlight key. The user will never explicitly type in the key, it will instead be provided by the UI.
```

or even more explicitly mentioning it will be the end-application/wallet